### PR TITLE
[CH-219] Support rlike/regexp_replace/regexp_extract/coalesce/DATE_ADD/DATE_SUB/DATEDIFF/modulo/concat_ws/collect_list/FROM_UNIXTIME

### PR DESCRIPTION
**Pull request by taiyang-li**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

test by CH[[230]]

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/expression/ArithmeticTransformer.scala [hunk 1]
```diff
@@ -284,4 +313,14 @@ object BinaryArithmeticTransformer {
...
     }
   }
+
+  def createRemainder(left: Expression, right: Expression,
+                   failOnError: Boolean, original: Expression): Expression = {
+    original match {
+      case r: Remainder =>
+        new RemainderTransformer(left, right, failOnError, r)
+      case other =>
+        throw new UnsupportedOperationException(s"not currently supported: $other.")
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/expression/ArithmeticTransformer.scala:323]: It seems using `UnsupportedOperationException` is not proper. `UnsupportedOperationException` is helping us let some unsupported operator/expression correctly fall back through catch statement. Here, I think it is developer's mistake in calling `createRemainder` for other kinds of expressions. Maybe, we can use `RuntimeException` or some other exception here. Right?

File: gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 2]
```diff
@@ -415,12 +415,14 @@ object ConverterUtils extends Logging {
...
   final val MAX = "max"
   final val STDDEV_SAMP = "stddev_samp"
+  final val COLLECT_LIST = "collect_list"
 
   // Function names used by Substrait plan.
   final val ADD = "add"
   final val SUBTRACT = "subtract"
   final val MULTIPLY = "multiply"
   final val DIVIDE = "divide"
+  final val REMAINDER = "remainder"
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:425]: It looks substrait is using `modulus` for remainder function. It would be better to align with substrait, right?

File: gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 3]
```diff
@@ -513,6 +515,10 @@ object ConverterUtils extends Logging {
 
   // SparkSQL DateTime functions of Velox
   final val DAY_OF_MONTH = "day_of_month"
+  final val FROM_UNIXTIME = "from_unixtime"
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:518]: The above comment may be misleading. If both backends share the same names for these functions, we can make a correction. Thanks!

File: gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 4]
```diff
@@ -513,6 +515,10 @@ object ConverterUtils extends Logging {
...
+  final val DATE_ADD = "date_add"
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:519]: We can align the name with substrait's for some of these functions. Right?
https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/extensions/functions_datetime.yaml#L37

File: gluten-core/src/main/java/io/glutenproject/substrait/expression/IntLiteralNode.java [hunk 5]
```diff
@@ -28,6 +28,10 @@ public IntLiteralNode(Integer value) {
     this.value = value;
   }
 
+  public Integer getValue() {
```
Comment (zzcclp) [inline gluten-core/src/main/java/io/glutenproject/substrait/expression/IntLiteralNode.java:31]: Why it needs to add this method for `IntLiteralNode` and `StringLiteralNode` ? For other xxxLiteralNode ?

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 6]
```diff
@@ -55,6 +59,51 @@ class BinaryExpressionTransformer(
...
 }
 
+
+class FromUnixTimeTransformer(sec: ExpressionTransformer, format: ExpressionTransformer,
+  timeZoneId: Option[String] = None, original: FromUnixTime)
+  extends ExpressionTransformer
+  with Logging {
+
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    if (BackendsApiManager.getBackendName.equalsIgnoreCase(GlutenConfig.GLUTEN_VELOX_BACKEND)) {
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:69]: add this expression to velox expression black list, don't check backend here.

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala [hunk 7]
```diff
@@ -56,6 +62,35 @@ class ExtractDateTransformer(
   }
 }
 
+
+class DateDiffTransformer(endDate: ExpressionTransformer, startDate: ExpressionTransformer,
+  original: DateDiff)
+  extends ExpressionTransformer with Logging {
+
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    if (BackendsApiManager.getBackendName.equalsIgnoreCase(GlutenConfig.GLUTEN_VELOX_BACKEND)) {
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:71]: add this expression to velox expression black list, don't check backend here.

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 8]
```diff
@@ -55,6 +59,51 @@ class BinaryExpressionTransformer(
...
+      throw new UnsupportedOperationException(s"not supported yet.")
+    }
+
+    val secNode = sec.doTransform(args)
+    var formatNode = format.doTransform(args)
+
+    // Only when format = 'yyyy-MM-dd HH:mm:ss' can we transfrom the expr to substrait.
+    if (!secNode.isInstanceOf[ExpressionNode] ||
+      !formatNode.isInstanceOf[StringLiteralNode] ||
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:78]: remove these two check

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala [hunk 9]
```diff
@@ -56,6 +62,35 @@ class ExtractDateTransformer(
...
+      throw new UnsupportedOperationException(s"not supported yet.")
+    }
+
+    // In Spark: datediff(endDate, startDate)
+    // In CH: date_diff('day', startDate, endDate)
+    val endDateNode = endDate.doTransform(args)
+    val startDateNode = startDate.doTransform(args)
+    val unitNode = ExpressionBuilder.makeStringLiteral("day")
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:79]: please check whether the velox backend need this parameter day

File: gluten-core/src/main/scala/io/glutenproject/expression/QuaternaryExpressionTransformer.scala [hunk 10]
```diff
@@ -24,6 +24,41 @@ import io.glutenproject.substrait.expression.{ExpressionBuilder, ExpressionNode}
...
+
+
+    val posNode = pos.doTransform(args)
+    if (!posNode.isInstanceOf[IntLiteralNode] ||
+      posNode.asInstanceOf[IntLiteralNode].getValue() != 1) {
+      throw new UnsupportedOperationException(s"${original} not supported yet.")
+    }
+
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    val functionName = ConverterUtils.makeFuncName(ExpressionMappings.REGEXP_REPLACE,
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/QuaternaryExpressionTransformer.scala:52]: use `substraitExprName: String` to pass substrait expression name, dont use `ExpressionMappings.REGEXP_REPLACE`

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 11]
```diff
@@ -55,6 +59,51 @@ class BinaryExpressionTransformer(
...
+    formatNode = ExpressionBuilder.makeStringLiteral("%Y-%m-%d %R:%S")
+
+    val dataTypes = if (timeZoneId != None) {
+      Seq(original.sec.dataType, original.format.dataType, StringType)
+    } else {
+      Seq(original.sec.dataType, original.format.dataType)
+    }
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    val functionId = ExpressionBuilder.newScalarFunction(functionMap,
+      ConverterUtils.makeFuncName(ExpressionMappings.FROM_UNIXTIME, dataTypes))
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:91]: use `substraitExprName: String` to pass substrait expression name, dont use `ExpressionMappings.FROM_UNIXTIME`, please refer to `HashExpressionTransformer`

File: gluten-core/src/main/scala/io/glutenproject/expression/ArithmeticTransformer.scala (hunk 1 above)
Comment (taiyang-li) [reply gluten-core/src/main/scala/io/glutenproject/expression/ArithmeticTransformer.scala]: RemainderTransformer is removed now.

File: gluten-core/src/main/java/io/glutenproject/substrait/expression/IntLiteralNode.java (hunk 5 above)
Comment (taiyang-li) [reply gluten-core/src/main/java/io/glutenproject/substrait/expression/IntLiteralNode.java:31]: It is used in RegExpReplaceTransformer. 
- First get value of pos by IntLiteralNode.getValue
- For clickhouse backend, if pos value is 1, then transformation is supported, otherwise throw exception. 
-

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala (hunk 9 above)
Comment (taiyang-li) [reply gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:79]: It is needed too: https://facebookincubator.github.io/velox/functions/datetime.html#date_diff

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 12]
```diff
@@ -18,12 +18,15 @@
 package io.glutenproject.expression
 
 import com.google.common.collect.Lists
-
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:21]: keep the blank line

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 13]
```diff
@@ -18,12 +18,15 @@
...
 import io.glutenproject.expression.ConverterUtils.FunctionConfig
 import io.glutenproject.substrait.expression.{ExpressionBuilder, ExpressionNode}
+import io.glutenproject.substrait.expression.StringLiteralNode
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:23]: revert these changes?

File: gluten-ut/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenArithmeticExpressionSuite.scala [hunk 14]
```diff
@@ -20,5 +20,4 @@ package org.apache.spark.sql.catalyst.expressions
 import org.apache.spark.sql.GlutenTestsTrait
 
 class GlutenArithmeticExpressionSuite extends ArithmeticExpressionSuite with GlutenTestsTrait {
-
```
Comment (zzcclp) [inline gluten-ut/src/test/scala/org/apache/spark/sql/catalyst/expressions/GlutenArithmeticExpressionSuite.scala:23]: remove this change ?

**Comment by zzcclp** [comment]
is there any ut covering these expressions?

File: gluten-core/src/main/scala/io/glutenproject/expression/QuaternaryExpressionTransformer.scala [hunk 15]
```diff
@@ -24,6 +24,42 @@ import io.glutenproject.substrait.expression.{ExpressionBuilder, ExpressionNode}
...
+  original: RegExpReplace)
+  extends ExpressionTransformer with Logging {
+
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    if (BackendsApiManager.getBackendName.equalsIgnoreCase(GlutenConfig.GLUTEN_VELOX_BACKEND)) {
+      QuaternaryExpressionTransformer(ExpressionMappings.REGEXP_REPLACE,
+        subject, regexp, rep, pos, original).doTransform(args)
+    }
+
+
```
Comment (lviiii) [inline gluten-core/src/main/scala/io/glutenproject/expression/QuaternaryExpressionTransformer.scala:45]: One empty line is enough, is there any other difference between QuaternaryExpressionTransformer and QuaternaryExpressionTransformer except posNode? Could you merge two into one?

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 16]
```diff
@@ -18,12 +18,14 @@
...
 import com.google.common.collect.Lists
-
 import io.glutenproject.expression.ConverterUtils.FunctionConfig
 import io.glutenproject.substrait.expression.{ExpressionBuilder, ExpressionNode}
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.BackendsApiManager
 
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.types.StringType
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:28]: revert these changes

**Review by PHILO-HE — APPROVED** [review]

**Review by zzcclp — APPROVED** [review]
LGTM
